flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/operators/sink/AbstractStreamingCommitterHandler.java

```diff
@@ -140,7 +142,23 @@ public void snapshotState(StateSnapshotContext context) throws Exception {
         }
 
         LOG.info("Committing the state for checkpoint {}", checkpointId);
-        recoveredCommittables(commit(readyCommittables));
+        List<CommT> failedCommittables = commit(readyCommittables);
+        recoveredCommittables(failedCommittables);
+
+        // Do not forward failed committables
+        final Set<CommT> failedCommittableSet = Collections.newSetFromMap(new IdentityHashMap<>());
+        failedCommittableSet.addAll(failedCommittables);
+
+        // Retain ordering of the committables
+        final Iterator<CommT> committableIt = readyCommittables.iterator();
+        while (committableIt.hasNext()) {
+            final CommT committable = committableIt.next();
+            if (!failedCommittableSet.contains(committable)) {
+                continue;
+            }
+            committableIt.remove();
```

Comment:
Is there a reason/benefit to removing the comittables from the original List instead of doing something like `return readyCommitables.stream.filter(...).collect(Collectors.toList())`?

Reply:
Not really I adjust it to use the stream notion.